sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java

```diff
@@ -127,28 +140,49 @@ public void writeThenReadAll() {
 
     PipelineResult result = pipeline.run();
     result.waitUntilFinish();
-    publishGcsResults(result);
+    gatherAndPublishMetrics(result);
   }
 
-  private void publishGcsResults(PipelineResult result) {
+  private void gatherAndPublishMetrics(PipelineResult result) {
+    String uuid = UUID.randomUUID().toString();
+    Timestamp timestamp = Timestamp.now();
+    List<NamedTestResult> namedTestResults = readMetrics(result, uuid, timestamp);
+    publishToBigQuery(namedTestResults, bigQueryDataset, bigQueryTable);
+    ConsoleResultPublisher.publish(namedTestResults, uuid, timestamp.toString());
+  }
+
+  private List<NamedTestResult> readMetrics(
+      PipelineResult result, String uuid, Timestamp timestamp) {
+    List<NamedTestResult> results = new ArrayList<>();
+
+    MetricsReader reader = new MetricsReader(result, FILEIOIT_NAMESPACE);
+    long writeStartTime = reader.getStartTimeMetric("startTime");
+    long writeEndTime = reader.getEndTimeMetric("middleTime");
+    long readStartTime = reader.getStartTimeMetric("middleTime");
+    long readEndTime = reader.getEndTimeMetric("endTime");
+    double writeTime = (writeEndTime - writeStartTime) / 1000.0;
+    double readTime = (readEndTime - readStartTime) / 1000.0;
+    double copiesPerSec = calculateGcsMetric(result);
+
+    if (copiesPerSec > 0) {
+      results.add(
+          NamedTestResult.create(uuid, timestamp.toString(), "copies_per_sec", copiesPerSec));
+    }
+
+    results.add(NamedTestResult.create(uuid, timestamp.toString(), "read_time", readTime));
+    results.add(NamedTestResult.create(uuid, timestamp.toString(), "write_time", writeTime));
+
+    return results;
+  }
+
+  private double calculateGcsMetric(PipelineResult result) {
```

Comment:
Could we call this `getGcsCopies` or `calculateGcsCopies` (I'm fine with both)?

Reply:
More importantly, this should be collected only when `options.getGcsPerformanceMetrics()` is `true`. I think this was missing before. Could you add that condition?

@udim am I right here or am I missing something?

Tl;dr version for Udi: we want to reuse the mechanism for publishing the bigQuery metrics for collecting time in IOITs too. Previously we checked if BigQuery related options are present to publish the Gcs metric. Now that there are more metrics collected this way, such "if" condition would be too broad.

Reply:
Added an option switching GCS metric reporting to file based tests

Reply:
I guess the way it was working before was by checking `if (numCopies < 0 || copyTimeMsec < 0)`, which is false if `options.getGcsPerformanceMetrics()` is false.
So I think we can merge --reportGcsPerformanceMetrics and --gcsPerformanceMetrics into the latter. No need to add a separate flag, and only report copiesPerSec if --gcsPerformanceMetrics is set.

Reply:
Thanks for the input! You're right, that would be optimal. However, I don't see a way to access this switch from inside the test - it's a part of `GcsOptions` class. Trying to use an option with the same name in two places leads to compilation problems (two clashing method definitions). Is there another, better option that you know of?

Reply:
Have you tried: `options.as(GcsOptions.class).getGcsPerformanceMetrics()`?

Reply:
I didn't try this, thanks for the suggestion! I changed the test so that the additional flag is not necessary.

Reply:
Interesting: if I read this correctly, we didn’t have to import the `sdks-java-extensions-google-cloud-platform` dependency because it is a transitive dependency from test-utils package. What if we refactor test utils and get rid of the extensions there? If we stick with one flag, we should import the `sdks-java-extensions-google-cloud-platform` dependency in `file-based-io-tests` module.

Looking at this in a broader, more future-proof way, I’ve begun to think it's better to stick with 2 separate flags: first flag in the production code that makes the IO/pipeline/whatever else actually produce the metric and the other one in the test code that explicitly tells that the metric should be harvested. IMO it has the following benefits: 

 - we separate the concerns of producing and collecting the metrics - people reading test code/implementing new tests can find/define the latter option in TestPipelineOptions interface, read it’s the description and decide if it should be used without digging into beam implementation specifics and searching for the option in the whole codebase 
 - we don't have to import runner/IO specific dependencies every time similar case occurs (like: ‘sdks-java-extensions-google-cloud-platform`) - this is especially important in case we’re running on other runners than Dataflow. I can imagine that tests run on Flink may not require dataflow/gcs related dependencies

The disadvantage is that we have 2 flags for each case like this but I think such solution will “scale” better in case we require more flags like this from other filesystems/runners (eg s3 copies/hdfs copies).

wdyt? 

Reply:
@lgajowy I think you're right that separating the setting of collection and reporting makes more sense, especially considering build dependencies.
For metric reporting, a single flag in the style of --experiments might simplify things (no need to add a new flag every time a metric is added).

Reply:
@udim What about cases in which we need some metrics, but not others? Take for example the hdfs copies, when we want this metric, other may be irrelevant, like gs copies. In that case we'd need to check if each metric has valid values and then decide whether to report it (like it's done now with gcs copies - number of operations greater than 0 and time not below 0? Calculate and report). I believe setting explicit flags for every metric wouldn't be simpler, but it would be easier to precisely control what is gathered and reported.

Reply:
@mwalenia Explicit flags for each metric isn't simpler, but perhaps using just 2 flags:
```
--metricsGather=hdfsCopies,gcsCopies,...
--metricsReportToBigQuery=hdfsCopies,gcsCopies,...
```

Reply:
@udim A big advantage of using my approach is built-in error reporting and checking. After passing an argument with a typo we get for example:

`java.lang.IllegalArgumentException: Class interface org.apache.beam.sdk.testing.TestPipelineOptions missing a property named 'compresionType'. Did you mean 'compressionType'?`

whereas there's no such mechanism when passing a list of strings as parameter value. In case of a typo we'd have a silent failure, since in order to switch reporting of different metrics on and off, we'd need to search for strings in an array.

I think we should stick with explicitly set flags for subsequent metrics. WDYT?

Reply:
Sounds good!